### PR TITLE
Add support for emscripten 1.38.46 and begin WASI support.

### PR DESCRIPTION
**Pull request by jplevyak**
Signed-off-by: John Plevyak <jplevyak@gmail.com>

### CODE REVIEW

File: test/extensions/wasm/wasm_test.cc [hunk 1]
```diff
@@ -175,10 +175,10 @@ TEST_P(WasmTest, EmscriptenVersion) {
   uint32_t major = 9, minor = 9, abi_major = 9, abi_minor = 9;
   EXPECT_TRUE(wasm->getEmscriptenVersion(&major, &minor, &abi_major, &abi_minor));
   EXPECT_EQ(major, 0);
-  EXPECT_LE(minor, 2);
+  EXPECT_LE(minor, 3);
```
Comment (PiotrSikora) [inline test/extensions/wasm/wasm_test.cc]: Could you add support for the 0.3 metadata version (i.e. extract the new metadata value)?

**Comment by PiotrSikora** [comment]
Feel free to ignore `STANDALONE_WASM=1` request (I filed #212 for it).

**Comment by PiotrSikora** [comment]
Ah, 1.38.46 requires `env.___wasi_fd_write`, so please ignore my request to split this as well.

File: api/wasm/cpp/Makefile.base [hunk 2]
```diff
@@ -11,7 +11,7 @@ else
 endif
 
 %.wasm %.wat: %.cc ${CPP_API}/proxy_wasm_intrinsics.h ${CPP_API}/proxy_wasm_enums.h ${CPP_API}/proxy_wasm_externs.h ${CPP_API}/proxy_wasm_api.h ${CPP_API}/proxy_wasm_intrinsics.js ${CPP_CONTEXT_LIB}
-	em++ -s WASM=1 -s BINARYEN_TRAP_MODE='clamp' -s LEGALIZE_JS_FFI=0 -s EMIT_EMSCRIPTEN_METADATA=1 --std=c++17 -O3 -g3 -I${CPP_API} -I/usr/local/include --js-library ${CPP_API}/proxy_wasm_intrinsics.js $*.cc ${CPP_API}/proxy_wasm_intrinsics.pb.cc ${CPP_CONTEXT_LIB} ${CPP_API}/libprotobuf.bc -o $*.js
+	em++ -s WASM=1 -s LEGALIZE_JS_FFI=0 -s EMIT_EMSCRIPTEN_METADATA=1 --std=c++17 -O3 -g3 -I${CPP_API} -I/usr/local/include --js-library ${CPP_API}/proxy_wasm_intrinsics.js $*.cc ${CPP_API}/proxy_wasm_intrinsics.pb.cc ${CPP_CONTEXT_LIB} ${CPP_API}/libprotobuf.bc -o $*.js
```
Comment (PiotrSikora) [inline api/wasm/cpp/Makefile.base]: Removing `-s BINARYEN_TRAP_MODE='clamp'` breaks `WasmTest.DivByZero` test for both runtimes (after regenerating `.wasm` modules), and it doesn't seem to be related to this change. Please revert it.

Comment (jplevyak) [reply api/wasm/cpp/Makefile.base]: Done.

File: source/extensions/common/wasm/wasm.cc [hunk 3]
```diff
@@ -892,6 +892,59 @@ Word ___syscall146Handler(void* raw_context, Word, Word syscall_args_ptr) {
...
+    return 8; // __WASI_EBADF
+  }
+
+  std::string s;
+  for (size_t i = 0; i < iovs_len.u64_; i++) {
+    auto memslice =
+        context->wasmVm()->getMemory(iovs.u64_ + i * 2 * sizeof(uint32_t), 2 * sizeof(uint32_t));
+    if (!memslice) {
+      context->wasm()->setErrno(EINVAL);
+      return -1;
```
[the remark this answers is not in the document]
Comment (jplevyak) [reply source/extensions/common/wasm/wasm.cc]: Done.

File: test/extensions/wasm/wasm_test.cc (hunk 1 above)
Comment (jplevyak) [reply test/extensions/wasm/wasm_test.cc]: Done.

File: source/extensions/common/wasm/wasm.cc [hunk 4]
```diff
@@ -892,6 +892,59 @@ Word ___syscall146Handler(void* raw_context, Word, Word syscall_args_ptr) {
   return written;
 }
 
+// Implementation of _wasi_fd_write() syscall that redirects stdout/stderr to Envoy logs.
+// _was_errno_t _wasi_fd_write(_wasi_fd_t fd, const _wasi_ciovec_t *iov, size_t iovs_len, size_t*
+// nwritten);
+Word ___wasi_fd_writeHandler(void* raw_context, Word fd, Word iovs, Word iovs_len,
```
[the remark this answers is not in the document]
Comment (jplevyak) [reply source/extensions/common/wasm/wasm.cc]: Done.
